File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        LinkedList<Integer>st=new LinkedList<>();
        st.addFirst(-1);
        int n=heights.length;
        
        int maxArea=0;
        
        for(int i=0;i<=n;i++){
            int val = i==n?0:heights[i];
            
            while(st.getFirst()!=-1 && val<=heights[st.peek()]){
                int rm=i;
                
                int h=heights[st.removeFirst()];
                int lm=st.peek();
                
                maxArea=Math.max(maxArea,h*(rm-lm-1));
                
            }
          st.addFirst(i);
        }

        return maxArea;
    }
}

File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int>st;
        int n=heights.size();
        int maxArea=0;
        
        for(int i=0;i<=n;i++){
            while(!st.empty() && (i==n || heights[st.top()]>=heights[i])){
                
                int h=heights[st.top()];
                st.pop();
                
                int w;
                if(st.empty())w=i;
                else w=i-st.top()-1;
                
                maxArea=max(maxArea,w*h);
            }
            st.push(i);
        }

        return maxArea;
    }
};